plugins/net_plugin/net_plugin.cpp

```diff
@@ -1097,32 +1100,38 @@ namespace eosio {
       if( ec ) {
          fc_elog( logger, "connection failed (set_option) ${peer}: ${e1}", ("peer", peer_name())( "e1", ec.message() ) );
          close();
-         return false;
+         callback(false);
       } else {
 
-         auto start_read = [this](){
-            fc_dlog( logger, "connected to ${peer}", ("peer", peer_name()) );
-            socket_open = true;
-            start_read_message();
-         };
          if (my_impl->ssl_enabled) {
             socket.ssl_socket()->async_handshake(server ? ssl::stream_base::server : ssl::stream_base::client,
-                                       boost::asio::bind_executor(strand, [start_read, c = shared_from_this(), socket=socket](const auto& ec){
+                                       boost::asio::bind_executor(strand, 
+                                       [c = shared_from_this(), socket=socket, callback](const auto& ec){
                                           //we use socket just to retain connection shared_ptr just in case it will be deleted
                                           //when we will need it inside start_read_message
                                           std::ignore = socket;
                                           if (ec) {
+                                             c->set_participating(false);
                                              fc_elog( logger, "ssl handshake error: ${e}", ("e", ec.message()) );
                                              c->close();
                                              return;
                                           }
-                                          start_read();
+
+                                          c->set_socket_open();
+                                          c->set_participating(true);
```

Comment:
I moved this code from `verify_certificate` for a reason. inside `verify_certificate` we do not guarantee that handshake was successful and we do not know if asio changed all internal structures already. So I moved it to later stage just because `socket_open` boolean and participating information can be used asynchronously.